### ISSUE TITLE
Refactor QueryRunner.run(), make responseContext immutable

### ISSUE THREAD

**Issue by leventov** [apache/incubator-druid#4113]
Mutable Map, passed into QueryRunner.run(), is abused: https://github.com/druid-io/druid/issues/4112 and inherently broken with async/concurrent runners: https://github.com/druid-io/druid/issues/3803.

I suggest to change QueryRunner signature to
```java
  QueryResult<T> run(Query<T> query);
```

where
```java
class QueryResult<T>
{
  final Sequence<T> result;
  final ImmutableMap<String, Object> responseContext;
}
```

What do you think? @himanshug @gianm

**Comment by gianm** [comment]
I guess that means the response context would need to be returned before queries are finished executing, since they won't be done by the time QueryRunner.run returns. Is that ok?

Example: at one point I was thinking we could use the response context for a possible query profiling feature, something that doesn't involve emitting metrics but instead returns a profile in the response itself. That would probably be tough to do with the change you proposed here, since not all profiling info is known at the time QueryRunner.run returns. I don't plan to work on this feature in the near future though, it was just something I thought might be cool.

Also, I think the fact that CachingClusteredClient merges sequences from servers would mean it would have to know how to merge response contexts too.

**Comment by gianm** [comment]
#3803 does mean that response contexts are not very useful right now, so I agree _something_ needs to change.

**Comment by gianm** [comment]
Would it be too weird to have:

```
class QueryResult<T>
{
  final Sequence<T> result;
  final Future<ImmutableMap<String, Object>> responseContext;
}
```

Where `responseContext` is a future that might not be available until the `result` has finished accumulating?

**Comment by gianm** [comment]
@leventov in #3803 you said you were using response contexts for more things -- what were they? Some use cases would illuminate what the behavior should be. Right now, in Druid response contexts aren't used for very much so it's not really clear to me what the best behavior is.

**Comment by leventov** [comment]
> I guess that means the response context would need to be returned before queries are finished executing, since they won't be done by the time QueryRunner.run returns. Is that ok?

Yes, I forgot that sometimes responseContext is updated *during* the Sequence accumulating. Also AFAIR sometimes it is somehow used *before* sequence is accumulated. So either we need to rethink some logic using responseContext now, or have two responseContexts: "before" and "after", that would be too complex IMO.

In #3803 I was using it to pass what is now known as `QueryMetrics`. However I should have used Query context for this, or a separate `QueryMetrics` property to `Query`. So I was abusing `responseContext` as well as in #4112.

**Comment by himanshug** [comment]
#3803 definitely needs to be fixed 

responseContext has very lightweight usage. one option is to change to `QueryRunner.run(Query query, ResponseContext context)` and ResponseContext class takes care of concurrency where necessary, this might be more straightforward than having 

```
QueryResult {
   ImmutableMap preContext;
   Sequence result;
   Future<ImmutableMap> postContext;
}
```


@gianm I have tried to do the "profiling" stuff in https://github.com/druid-io/druid/pull/3371

**Comment by leventov** [comment]
@himanshug so we are suggesting the same thing. I raised concern that this design is overly complex, but it might be unavoidable, if we genuinely need preContext and postContext sometimes.

**Comment by gianm** [comment]
IMO, changes to response context should be motivated by use cases. There is no need to add complexity for things we might need but don't actually need. I suggest we combine this issue with #3803, they seem to be the same underlying problem.

Right now, response context is used for:

- tracking missing segments for query retries (see SpecificSegmentQueryRunner, ReportTimelineMissingSegmentQueryRunner, RetryQueryRunner)
- returning a list of `uncoveredIntervals` to the client (see #2058, made optional in #2332)
- tracking timeouts in the Scan query (although this usage is a bit dubious, see #4112)

That's all I can think of, maybe there's more that I missed. Possible future usages include query profiling: https://github.com/druid-io/druid/issues/4113#issuecomment-289181050 and #3371.

Anyone aware of anything else?

**Comment by himanshug** [comment]
response context is also used for tracking the "etag" field to implement https://github.com/druid-io/druid/pull/3955

**Comment by leventov** [comment]
@gianm the motivation for change is that it's impossible to reason about correctness with the current design. Probably it's ok right now, but probably not OK already, we may silently miss some queryContext entries and never notice that. It also may depend on the cluster deployment scheme. E. g. if brokers and historicals are deployed on the same nodes.

Generally, shared mutable response context breaks QueryRunner isolation, some combinations of QueryRunners work, but some will definitely not.

I support the following design:
```
QueryResult {
   ImmutableMap preResponseContext;
   Sequence result;
   Future<ImmutableMap> postResponseContext;
}
```

IMO the change should be combined with #4131.

**Comment by leventov** [comment]
Addressed by #8110
